yuyo/repeaters.py

```diff
@@ -0,0 +1,228 @@
+
+_function = Callable[..., Awaitable[Any]]
+RepeaterFuncT = TypeVar("RepeaterFuncT", bound=_function)
+
+
+class Repeater(Generic[RepeaterFuncT]):
+    def __init__(
+        self,
+        coro: RepeaterFuncT,
+        *,
+        seconds: Optional[float] = None,
+        minutes: Optional[float] = None,
+        hours: Optional[float] = None,
+        run_count: Optional[int] = None,
+        loop: Optional[asyncio.AbstractEventLoop] = None,
+    ):
+        if not (seconds or minutes or hours):
+            raise  # TODO: What error should this raise?
+        self._delay: float = (hours or 0) * 3600 + (minutes or 0) * 60 + (seconds or 0)
+        self._runs_left = run_count or -1
+        self._event_loop = loop or asyncio.get_event_loop()
+        self._coro = coro
+        self._task: Optional[asyncio.Task[None]] = None
+        self._iter: int = 0
+        self._next_iteration: Optional[datetime.datetime] = None
+        self._before: Optional[RepeaterFuncT] = None
+        self._after: Optional[RepeaterFuncT] = None
+        self.ignored_exceptions: List[type] = []
+        self.fatal_exceptions: List[type] = []
+
+    async def _wrapped_coro(self):
+        try:
+            await self._coro()
+        except BaseException as e:  # noqa - I have to
+            if type(e) in self.fatal_exceptions:
+                self.stop()
+                raise
+            if type(e) not in self.ignored_exceptions:
+                raise
+
+    async def _loop(self):
+        if self._before:
+            await self._before()
+        while self._runs_left != 0:
+            self._iter += 1
+            self._runs_left -= 1
+            self._event_loop.create_task(self._wrapped_coro())
+            await asyncio.sleep(self._delay)
+        if self._after:
+            await self._after()
+
+    @property
+    def iteration_count(self) -> int:
+        """Return the iteration this repeater is on."""
+        return self._iter
+
+    def start(self) -> asyncio.Task:
+        """
+        Start the repeater.
+
+        Returns
+        -------
+        :class:`asyncio.Task`
+            The started task
+        """
+        if self._task is not None and not self._task.done():
+            raise RuntimeError("Repeater already running")
+        self._task = self._event_loop.create_task(self._loop())
+        return self._task
+
+    def stop(self):
+        """Cancel the repeater."""
+        if self._task is None or self._task.done():
+            raise RuntimeError("Repeater not running")
+        self._task.cancel()
+        if self._after:
+            self._event_loop.create_task(self._after())
+
+    def with_pre_callback(self, coro: RepeaterFuncT):
+        if not inspect.iscoroutinefunction(coro):
+            raise TypeError(f"Pre callback must be a coroutine, got {coro.__class__.__name__}.")
+        self._before = coro
+
+    def with_post_callback(self, coro: RepeaterFuncT):
+        if not inspect.iscoroutinefunction(coro):
```

Comment:
I personally prefer to avoid this check cause there's cases it ignores like callable objects